File: pydroid/src/main/java/com/pyamsoft/pydroid/about/AboutLibrariesPresenter.java
/*
 * Copyright 2017 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.pydroid.about;

import android.support.annotation.NonNull;
import com.pyamsoft.pydroid.presenter.SchedulerPresenter;
import io.reactivex.Scheduler;
import timber.log.Timber;

public class AboutLibrariesPresenter extends SchedulerPresenter {

  @NonNull private final AboutLibrariesInteractor interactor;

  @SuppressWarnings("WeakerAccess")
  public AboutLibrariesPresenter(@NonNull AboutLibrariesInteractor interactor,
      @NonNull Scheduler observeScheduler, @NonNull Scheduler subscribeScheduler) {
    super(observeScheduler, subscribeScheduler);
    this.interactor = interactor;
  }

  public void loadLicenses(@NonNull LoadCallback callback) {
    disposeOnStop(interactor.loadLicenses()
        .subscribeOn(getSubscribeScheduler())
        .observeOn(getObserveScheduler())
        .subscribe(callback::onLicenseLoaded,
            throwable -> Timber.e(throwable, "onError loading licenses")));
  }

  public interface LoadCallback {

    void onLicenseLoaded(@NonNull AboutLibrariesModel model);
  }
}

File: pydroid/src/main/java/com/pyamsoft/pydroid/about/AboutLibrariesPresenter.kt
/*
 *     Copyright (C) 2017 Peter Kenji Yamanaka
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License along
 *     with this program; if not, write to the Free Software Foundation, Inc.,
 *     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

package com.pyamsoft.pydroid.about

import com.pyamsoft.pydroid.about.AboutLibrariesPresenter.LoadCallback
import com.pyamsoft.pydroid.presenter.SchedulerPresenter
import io.reactivex.Scheduler
import timber.log.Timber

class AboutLibrariesPresenter internal constructor(private val interactor: AboutLibrariesInteractor,
    computationScheduler: Scheduler, ioScheduler: Scheduler,
    mainThreadScheduler: Scheduler) : SchedulerPresenter<LoadCallback>(computationScheduler,
    ioScheduler, mainThreadScheduler) {

  override fun onBind(v: LoadCallback) {
    super.onBind(v)
    loadLicenses(false, v::onLicenseLoaded, v::onAllLoaded)
  }

  private fun loadLicenses(force: Boolean, onLicenseLoaded: (AboutLibrariesModel) -> Unit,
      onAllLoaded: () -> Unit) {
    dispose {
      interactor.loadLicenses(force).subscribeOn(ioScheduler).observeOn(
          mainThreadScheduler).doAfterTerminate { onAllLoaded() }.subscribe({ onLicenseLoaded(it) },
          { Timber.e(it, "onError loading licenses") })
    }
  }

  interface LoadCallback {

    /**
     * Called when a single license has finished loading. There are no guarantees about if the
     * license was loaded for the first time.
     */
    fun onLicenseLoaded(model: AboutLibrariesModel)

    /**
     * Called when all licenses are done loading
     */
    fun onAllLoaded()
  }
}
